docs/references/ingress-messages.mdx

```diff
@@ -0,0 +1,84 @@
+   1. The node is offline.
+   2. An execution pre-processing error (status `200` in case of the async call) occurs.
+   3. The node is incapable of handling the request (status `5xx`).
+4. It could be that status responses are never received. For example, a malicious boundary node could block the status response. Thus, the agent would never see a [status like `received`, `processing`, etc.](/docs/current/references/ic-interface-spec#state-tree-request-status) upon calling `read_state`.
+5. The certificate in the `read_state` response could be invalid. For example, it could be malformed or have an invalid signature.
+6. Finally, the call could be `rejected` (see [rejected status](/docs/current/references/ic-interface-spec#state-tree-request-status) and [reject codes](/docs/current/references/ic-interface-spec#reject-codes)).
+
+When seeing errors 2-5, the call may still be successfully executed. For example, the user's might lose network connectivity, or a malicious boundary node or node could return an error but still successfully execute the request. Only if a `rejected` response is received along with a valid certificate, the subnet guarantees that the call hasn’t been executed and never will be.
+
+Some operations, such as transfers, must not be executed multiple times. If a call was `rejected` as described above, it is safe to retry the operation. In all other error cases however, retrying can be risky because the call could still successfully execute, potentially leading to severe security vulnerabilities such as double spending.
+
+In general, there are cases where it is impossible to determine if a call was successful. For example, suppose
+* The client loses its connection until the request status has been removed from the state tree. (Recall that ICP will remove the request from the system state tree some time after ingress expiry.)
+* There is no application specific way to tell if the call successfully executed. For example, the canister does not provide a way to find out if the call happened.
+
+In this situation, it is indistinguishable if the call was successful, but all information has already been removed from the system state tree, or the call never made it into the state tree in the first place and hasn't been executed. 
+
+### Recommendations and best practices for secure update call handling
+
+The best way to handle the aforementioned errors is to implement safe retries using idempotent calls. Idempotency and several patterns to achieve it are discussed in the [best practices on safe retries and idempotency](/docs/current/developer-docs/smart-contracts/best-practices/idempotency).
+
+It may not always be possible to make call endpoints idempotent. For example, an endpoint that is not under the developer’s control may simply not have this property. In this case, we still suggest retrying as a best effort. However, the guarantees in this case are weaker compared to using idempotent calls.
```

Comment:
WDYM with "retrying as a best effort"?
If the endpoint can't be made idempotent, we still advise to retry which might lead to vulnerabilities?

Reply:
Good point, this is misleading. What I meant is essentially what is described below, i.e. retrying with the same request ID. I changed the wording here to explain this better, PTAL!